### PR TITLE
Separate `executor` concept from `application` to support complex task scheduling and execution

### PR DESCRIPTION
**Pull request by greg7mdp**
For the parallelization of readonly transactions, we need to have more complex scheduling of tasks (for example separate queues for readonly and read/write transactions), and be able to update how tasks are scheduled depending on which type of transaction we want to process.

The previous architecture of appbase did not allow to customize the enqueuing and execution of tasks.

This PR introduce the notion of an `executor` class, which can be provided by an application in order to customize the enqueuing and processing of tasks. A `default_executor`  class is also provided which implements the same functionality as existed before.

This PR does not change the default appbase functionality or interface. `Nodeos` still builds with this version of appbase without any change, and all the tests pass.

This work was made in collaboration with @linh2931. He will submit a further PR adding more tests reflecting how appbase will be used in nodeos for parallelizing readonly transactions.

### CODE REVIEW

**Review by linh2931 — APPROVED** [review]
I have tested this. It worked. Will create a new PR with changes required for read-only transaction execution parallelization and tests.

**Comment by linh2931** [comment]
Got this when building nodeos. I am using the latest of the branch.

```
In file included from /home/lh/work/leap-appbase-test/tests/trx_generator/trx_generator.cpp:1:
In file included from /home/lh/work/leap-appbase-test/tests/trx_generator/trx_generator.hpp:5:
/home/lh/work/leap-appbase-test/plugins/chain_plugin/../../libraries/appbase/include/appbase/plugin.hpp:33:21: error: member access into incomplete type 'appbase::application'
               app().plugin_initialized(*this);
                    ^
/home/lh/work/leap-appbase-test/plugins/chain_plugin/../../libraries/appbase/include/appbase/plugin.hpp:6:10: note: forward declaration of 'appbase::application'
   class application;
         ^
/home/lh/work/leap-appbase-test/plugins/chain_plugin/../../libraries/appbase/include/appbase/plugin.hpp:46:21: error: member access into incomplete type 'appbase::application'
               app().plugin_started(*this);
                    ^
/home/lh/work/leap-appbase-test/plugins/chain_plugin/../../libraries/appbase/include/appbase/plugin.hpp:6:10: note: forward declaration of 'appbase::application'
   class application;
         ^
```

**Review by linh2931 — CHANGES_REQUESTED** [review]
Please fix the error causing following failure:

```
In file included from /home/lh/work/leap-appbase-test/tests/trx_generator/trx_generator.cpp:1:
In file included from /home/lh/work/leap-appbase-test/tests/trx_generator/trx_generator.hpp:5:
/home/lh/work/leap-appbase-test/plugins/chain_plugin/../../libraries/appbase/include/appbase/plugin.hpp:33:21: error: member access into incomplete type 'appbase::application'
               app().plugin_initialized(*this);
                    ^
/home/lh/work/leap-appbase-test/plugins/chain_plugin/../../libraries/appbase/include/appbase/plugin.hpp:6:10: note: forward declaration of 'appbase::application'
   class application;
         ^
/home/lh/work/leap-appbase-test/plugins/chain_plugin/../../libraries/appbase/include/appbase/plugin.hpp:46:21: error: member access into incomplete type 'appbase::application'
               app().plugin_started(*this);
```

**Comment by greg7mdp** [comment]
> Please fix the error causing following failure:

That's weird, I'm using clang and I only get warnings for this error.

I'm not quite sure why we didn't have this issue before my latest changes, but it is a problem in `tests/trx_generator/trx_generator.hpp` which includes `plugin.hpp` instead of `application.hpp` like we do everywhere else.

When we update the leap repo to point to the new appbase, we can fix that at the same time.

**Comment by greg7mdp** [comment]
Actually, because the file plugin.hpp is inconsistent on its own (it needs the `application` definition), I have removed it. There is an error when building leap, but it is easily resolved by removing #include <appbase/plugin.hpp> from tests/trx_generator/trx_generator.hpp , as it is not needed.

**Comment by linh2931** [comment]
> 

The clang version I am using is 15.

**Review by linh2931 — APPROVED** [review]

File: include/appbase/application.hpp [hunk 1]
```diff
@@ -1,383 +1,24 @@
...
+// ---------------------------------------------------------------------------------------------
+// to use a different executor:
+//
+//   1. create a copy of this file in your own application
+//
+//   2. create your own version of appbase/default_executor.hpp
+//           (it needs to define a type `appbase_executor`)
+//      and include this file instead of <appbase/default_executor.hpp>
+//
+//   3. include your own `application.hpp` in your project
```
Comment (heifner) [inline include/appbase/application.hpp:10]: This seems like too much of a hack to me. 

Seems like a user of appbase should choose which executor type either by a #define or better yet a type declaration before including `application.hpp`. I don't think we want users to have to hack up their copy of `appbase` to make it work.

Comment (heifner) [reply include/appbase/application.hpp:10]: How to use `appbase`:

- Include the executor type you want (or write your own).
- Declare your executor as the one to use:
   ```
      #include <appbase/default_executor.hpp>
      namespace appbase { using appbase_executor = appbase::default_executor; }
      #include <appbase/application.hpp>
   ```

File: include/appbase/default_executor.hpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+   private:
+      // members are ordered taking into account that the last one is destructed first
+      boost::asio::io_service  io_serv;
+      execution_priority_queue pri_queue;
+   };
+
+}
+
+using appbase_executor = appbase::default_executor;
```
Comment (heifner) [inline include/appbase/default_executor.hpp:42]: I think this using statement should be in `namespace appbase`

File: include/appbase/application_base.hpp [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+          */
+         void set_thread_priority_max();
+
+      void set_stop_executor_cb(std::function<void()> cb) {
+         stop_executor_cb = std::move(cb);
+      }
+
+      void set_post_cb(std::function<void (int, std::function<void()>)> cb) {
+         post_cb = std::move(cb);
+      }
```
Comment (heifner) [inline include/appbase/application_base.hpp:280]: These two are not indented like the others.
I don't think either one of these methods is needed.

File: include/appbase/application_base.hpp [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+          * Post func to run on io_service with given priority.
+          *
+          * @param priority can be appbase::priority::* constants or any int, larger ints run first
+          * @param func function to run on io_service
+          * @return result of boost::asio::post
+          */
+         template <typename Executor, typename Func>
+         auto post( Executor &exec, int priority, Func&& func ) {
+            return exec.post(priority, std::forward<Func>(func));
+         }
```
Comment (heifner) [inline include/appbase/application_base.hpp:264]: Seems like we will want to remove or at least deprecate this. I would vote for just removing it even though it would be a breaking change. Someone can just not update their `appbase` submodule until they are ready to use the new interface. The new interface will be:
`app().executor().post(...)` where `...` I assume will include some read-only vs write enum argument.

File: application_base.cpp [hunk 5]
```diff
@@ -191,19 +191,19 @@ void application::startup() {
...
 #endif
 }
 
-void application::start_sighup_handler( std::shared_ptr<boost::asio::signal_set> sighup_set ) {
+void application_base::start_sighup_handler( std::shared_ptr<boost::asio::signal_set> sighup_set ) {
 #ifdef SIGHUP
    sighup_set->async_wait([sighup_set, this](const boost::system::error_code& err, int /*num*/) {
       if( err ) return;
-      app().post(priority::medium, [sighup_set, this]() {
+      post_cb(priority::medium, [sighup_set, this]() {
```
Comment (heifner) [inline application_base.cpp:206]: Why was this changed? `app().post` still works here. Although, I would argue it should be changed to `app().executor().post`

File: application_base.cpp [hunk 6]
```diff
@@ -464,16 +455,16 @@ void application::shutdown() {
       std::rethrow_exception(eptr);
 }
 
-void application::quit() {
+void application_base::quit() {
    my->_is_quiting = true;
-   io_serv.stop();
+   stop_executor_cb();
```
Comment (heifner) [inline application_base.cpp:460]: This can be `app().get_io_service().stop()`

File: include/appbase/application_instance.hpp [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+      }
+
+      static void reset_app_singleton() { app_instance.reset(); }
+      
+      static bool null_app_singleton()  { return !app_instance; }
+
+      template <typename Func>
+      auto post( int priority, Func&& func ) {
+         return application_base::post(executor_, priority, std::forward<Func>(func));
+      }
```
Comment (heifner) [inline include/appbase/application_instance.hpp:24]: I think this should be removed (or at least deprecated) since the interface will be different for multi-queue posting.

File: include/appbase/application_instance.hpp [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+         application_base::exec(executor_);
+      }
+
+      boost::asio::io_service& get_io_service() {
+         return executor_.get_io_service();
+      }
+
+      auto& get_priority_queue() {
+         return executor_.get_priority_queue();
+      }
```
Comment (heifner) [inline include/appbase/application_instance.hpp:36]: Seems like these are not needed. Just let users interact with `app().executor()`

File: include/appbase/application.hpp (hunk 1 above)
Comment (greg7mdp) [reply include/appbase/application.hpp:10]: Thanks Kevin. In leap for example, this pattern would have to be present in multiple files, right (wherever `#include <appbase/application.h>` is currently found). I don't think it would be good to repeat this patterm, but that we should have a single header for it. My suggestion was something like `<leap/application.h>`, which was the reason of my suggestion.
What would you suggest we do in appbase? Add the 3 lines above wherever `#include <appbase/application.h>` is currently found)?

File: examples/executor_example/application.hpp [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+//
+//   1. create a copy of this file in your own application
+//
+//   2. create your own version of appbase/default_executor.hpp
+//           (it needs to define a type `appbase_executor`)
+//      and include this file instead of <appbase/default_executor.hpp>
+//
+//   3. include your own `application.hpp` in your project
+//
+// ---------------------------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply examples/executor_example/application.hpp:12]: Yes, the intent was to have a different executor from the default. I guess I need to finish that. I think it is good to have an example, so that  users don't have to figure things out from the test how to use a different executor. 
What would you prefer, that I finish the exeample with a different executor, or that I remove the example alltogether?

File: include/appbase/application.hpp (hunk 1 above)
Comment (heifner) [reply include/appbase/application.hpp:10]: `appbase` is currently only used by plugins. So an `application.hpp` in `plugins/chain_interface/include/eosio/chain/` is likely a good place for it.

Comment (heifner) [reply include/appbase/application.hpp:10]: > What would you suggest we do in appbase? Add the 3 lines above wherever #include <appbase/application.h> is currently found)?

Is there a problem with just making `application` a template? Then users could just do something like:
``` 
   #include <appbase/default_executor.hpp>
   #include <appbase/application.hpp>
   using app = appbase::application<default_executor>;
```

File: include/appbase/application_base.hpp (hunk 3 above)
Comment (greg7mdp) [reply include/appbase/application_base.hpp:280]: I will fix the indentation. These two methods are needed and used in the `application` constructor (application_instance.hpp:42`), as they indirectly provide access to the executor from inside `application_base` member functions (see `application_base.cpp`) who doesn't know about it as it is a member of `application.hpp`).

File: include/appbase/application.hpp (hunk 1 above)
Comment (greg7mdp) [reply include/appbase/application.hpp:10]: Both Lin and I looked at making the whole `application` a template (that was the first idea), but that would have required moving all the functions and the includes that are currently inside `application_base.cpp` into the header file, and we both felt that it was not a good idea and that the current solution is much preferable.

File: include/appbase/application_base.hpp (hunk 3 above)
Comment (greg7mdp) [reply include/appbase/application_base.hpp:229]: OK, the new header files are now consistently indented.

File: include/appbase/application.hpp (hunk 1 above)
Comment (heifner) [reply include/appbase/application.hpp:10]: You should be able to keep application_base.cpp as is and only make `application` a template.

File: include/appbase/application_base.hpp (hunk 4 above)
Comment (greg7mdp) [reply include/appbase/application_base.hpp]: done, it is removed.

File: application_base.cpp (hunk 5 above)
Comment (greg7mdp) [reply application_base.cpp:206]: agreed, ot should be `app().post()`.

It cannot be `app().executor().post` as the appbase user may change the executor() and we don't want the executor type hardcoded in appbase.

File: application_base.cpp (hunk 6 above)
Comment (greg7mdp) [reply application_base.cpp:460]: No, same as above, the appbase user may change the executor() and we don't want the executor type hardcoded in appbase, no API from `application` should be used in this file.

File: application_base.cpp (hunk 5 above)
Comment (heifner) [reply application_base.cpp:206]: I'm starting to think we should divorce executor from application. The executor could be provided to application where needed and users just interact with an executor outside of app()

File: include/appbase/application.hpp (hunk 1 above)
Comment (greg7mdp) [reply include/appbase/application.hpp:10]: > You should be able to keep application_base.cpp as is and only make `application` a template.

I don't see what problem that solves?

File: examples/executor_example/application.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (heifner) [reply examples/executor_example/application.hpp:12]: These concepts seem generic enough to me to be in appbase.

File: include/appbase/application_base.hpp (hunk 4 above)
Comment (linh2931) [reply include/appbase/application_base.hpp]: Probably we should deprecate it first before removing it.

File: include/appbase/application.hpp (hunk 1 above)
Comment (heifner) [reply include/appbase/application.hpp:10]: Maybe we should have a quick meeting. My main point is that this all seems like a hack. A user of `appbase` should be able to use it directly without having to modify anything in the `appbase` submodule. We should provide both a `default_executor` and a `multi-queue-executor` and users should be able to choose between them. Or we just make our fork of `appbase` only have a multi-queue-executor since that is all we need. Not sure we need to maintain the `default_executor`. Seems like you should be able to use the multi-queue one in a way that you just only use one of the queues giving you the current default behavior.

File: include/appbase/default_executor.hpp (hunk 2 above)
Comment (greg7mdp) [reply include/appbase/default_executor.hpp]: Thanks, done.

File: include/appbase/application_base.hpp (hunk 4 above)
Comment (heifner) [reply include/appbase/application_base.hpp]: We didn't deprecate anything when we added the `priority` option. I think we should add the new multi-queue option and not try to make it backwards compatible.

File: include/appbase/application.hpp (hunk 1 above)
Comment (greg7mdp) [reply include/appbase/application.hpp:10]: > A user of appbase should be able to use it directly without having to modify anything in the appbase submodule.

Yes, that is the case with the current implementation. It can be used without requiring any change in the appbase submodule.

Comment (greg7mdp) [reply include/appbase/application.hpp:10]: We can certainly have a quick meeting!

File: application_base.cpp (hunk 5 above)
Comment (greg7mdp) [reply application_base.cpp:206]: `application_base` uses the executor. If we have a separation, it will require a virtual interface for the executor which would be slower.

Comment (heifner) [reply application_base.cpp:206]: It could be passed into `exec()`.

File: include/appbase/application.hpp (hunk 1 above)
Comment (linh2931) [reply include/appbase/application.hpp:10]: We can provide a two-queue executor. The user can choose to post to either queue; by default the first queue is posted. We will have 2 execution modes: `from_first_queue` or `from_both_queues` and the user is able to set the mode either at the start or dynamically. I just thought those concepts were too specific for Leap. But now I think we should do this.

Comment (greg7mdp) [reply include/appbase/application.hpp:10]: Why would we do this? The whole point of the separation is that we want to allow the executor to be anything the user wants. Why hardcode two separate executors in appbase?

File: examples/executor_example/application.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (greg7mdp) [reply examples/executor_example/application.hpp:12]: We could provide it as the example maybe?

File: include/appbase/application.hpp (hunk 1 above)
Comment (heifner) [reply include/appbase/application.hpp:10]: I think we just change appbase to have only the new multi-queue executor. If you only need the one queue only post to the one queue.

File: application_base.cpp (hunk 5 above)
Comment (greg7mdp) [reply application_base.cpp:206]: The issue with this is that the user could pass an executor to exec(), and a different one to the other `application` APIs. I think the current implementation is safer, and I don't see the benefit of a change

File: include/appbase/application.hpp (hunk 1 above)
Comment (linh2931) [reply include/appbase/application.hpp:10]: I will abstract what I already have for Leap and submit a default two-queue executor (it requires a little more changes like managing the order among two queues and support multiple threading). Should be done soon after lunch.

File: application_base.cpp (hunk 5 above)
Comment (greg7mdp) [reply application_base.cpp:206]: But yes, maybe we should make `application` a template as you suggested, instead of passing the type in `executor_t`. Let me try that.

Comment (greg7mdp) [reply application_base.cpp:206]: @heifner I just updated `application` to be a template as you suggested.

File: include/appbase/application_instance.hpp (hunk 8 above)
Comment (greg7mdp) [reply include/appbase/application_instance.hpp]: They are needed because they are used in `application_base`.

File: include/appbase/application.hpp (hunk 1 above)
Comment (greg7mdp) [reply include/appbase/application.hpp]: > I think we just change appbase to have only the new multi-queue executor. 

I disagree. The executor used in leap will require more changes anyways, like having a thread_pool, etc... The point of this exercise was to *not* implement the needs of leap in appbase, so that leap can evolve without requiring changes to appbase.

Comment (linh2931) [reply include/appbase/application.hpp]: The users are free to implement their own executors. Appbase will just provide and maintain a default two-queue executor (thread-pool is not needed here). This two-queue executor still tries to be as generic as possible. Let me get it out and we will decide whether to maintain a one-queue or two-queue executor afterwards.

Comment (greg7mdp) [reply include/appbase/application.hpp]: I don't agree that  the two queue executor should be added in appbase, except maybe as an example. It adds a lot of unwarranted comp-lexity, and it is unlikely to be useful to any application except than leap. And as for leap, it will define its own executor anyways.

File: include/appbase/application_instance.hpp (hunk 7 above)
Comment (greg7mdp) [reply include/appbase/application_instance.hpp]: OK, I have deprecated it!

File: include/appbase/application_instance.hpp (hunk 8 above)
Comment (greg7mdp) [reply include/appbase/application_instance.hpp]: OK, I have removed it.

File: include/appbase/application.hpp (hunk 1 above)
Comment (greg7mdp) [reply include/appbase/application.hpp]: The way I see it, the `default_executor` and the `execution_priority_queue` are examples which should be as basic as possible. 

The `priority_queue` is only needed by the executor.  A complex application with specific needs would redefine its own `executor` and `priority_queue` (or whatever enqueueing mechanism needed by the executor).

File: include/appbase/application.hpp [hunk 10]
```diff
@@ -1,383 +1,23 @@
...
-         friend class plugin;
+// ---------------------------------------------------------------------------------------------
+// to use a different executor:
+//
+//   1. create a copy of this file in your own application
+//
+//   2. create your own executor type (see appbase/default_executor.hpp)
+//      and define an `application` type templated with it.
+//
+//   3. use as below
```
Comment (heifner) [inline include/appbase/application.hpp:11]: How about:
```
// --
// To use a different executor:
//
// 1. Create your own executor type  (see executor types already defined as examples).
// 2. Add a using statement indicating your application and include <appbase/application_instance.hpp>
//          namespace appbase {
//             using application = application_t<default_executor>;
//          }
//          #include <appbase/application_instance.hpp>
// --
```

File: tests/two_queue_tests.cpp [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+   // both queues are cleared after execution 
+   BOOST_REQUIRE_EQUAL( app->executor().queue1().empty(), true);
+   BOOST_REQUIRE_EQUAL( app->executor().queue2().empty(), true);
+
+   // exactly number of posts processed
+   BOOST_REQUIRE_EQUAL( rslts.size(), 6 );
+
+   // same priority (medidum) of functions executed by the post order
+   BOOST_CHECK_LT( rslts[0], rslts[1] );
```
Comment (heifner) [inline tests/two_queue_tests.cpp:91]: Comment and expected results are not correct.

File: include/appbase/application_base.hpp [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+      return executor_.get_io_service();
+   }
+
+   void startup() {
+      application_base::startup(get_io_service());
+   }
+
+   application_t() {
+      set_stop_executor_cb([&]() { get_io_service().stop(); });
+      set_post_cb([&](int prio, std::function<void()> cb) { executor_.post(prio, std::move(cb)); });
```
Comment (heifner) [inline include/appbase/application_base.hpp:372]: I don't see what this buys you over just using `executor_.post()` since you are hard-coding it here. It seems the only way this could be made to work generically is if you registered this callbacks before construction.

File: include/appbase/application_instance.hpp [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+   state _state = abstract_plugin::registered;
+   std::string _name;
+};
+
+// ------------------------------------------------------------------------------------------
+template <typename Data, typename DispatchPolicy>
+void channel<Data, DispatchPolicy>::publish(int priority, const Data& data) {
+   if (has_subscribers()) {
+      // this will copy data into the lambda
+      app().executor().post(priority, [this, data]() { _signal(data); });
```
Comment (heifner) [inline include/appbase/application_instance.hpp:69]: You are depending on the executor having a post that takes a priority and a lambda. So I think you can just require that and not need the `post_cb` std::function.

File: include/appbase/application.hpp (hunk 10 above)
Comment (greg7mdp) [reply include/appbase/application.hpp]: Sure, I have updated the comment, thank you @heifner .

File: include/appbase/application_base.hpp (hunk 12 above)
Comment (greg7mdp) [reply include/appbase/application_base.hpp:372]: That's because `application_base` uses `post()` and the executor is not defined yet, so we can't use `executor_` from application_base.

File: include/appbase/application_instance.hpp (hunk 13 above)
Comment (greg7mdp) [reply include/appbase/application_instance.hpp:69]: Same answer as above comment.

File: include/appbase/application_base.hpp (hunk 12 above)
Comment (greg7mdp) [reply include/appbase/application_base.hpp:372]: > Ah, yes, you changed where executor_ is stored.

Not since the PR was made, but yes before there was one class for `application`, now there is a base which has its regular members implemented in a .cpp file, and the `application` class itself which is a template class in a header.

File: tests/two_queue_tests.cpp [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+   app->executor().post( priority::low,    app->executor().queue1(), [&]() { rslts[4]=seq_num; ++seq_num; } );
+   app->executor().post( priority::low,    app->executor().queue2(), [&]() { rslts[5]=seq_num; ++seq_num; } );
+   app->executor().post( priority::highest,app->executor().queue1(), [&]() { rslts[6]=seq_num; ++seq_num; } );
+   app->executor().post( priority::high,   app->executor().queue2(), [&]() { rslts[7]=seq_num; ++seq_num; } );
+
+   // stop application. Use lowest at the end to make sure this executes the last
+   app->executor().post( priority::lowest, app->executor().queue1(), [&]() {
+      // queue1 should have current function and queue2 should have all its functions
+      BOOST_REQUIRE_EQUAL( app->executor().queue1().size(), 1);
+      BOOST_REQUIRE_EQUAL( app->executor().queue2().size(), 3 );
```
Comment (heifner) [inline tests/two_queue_tests.cpp:43]: BOOST TEST is not thread safe: https://www.boost.org/doc/libs/1_81_0/libs/test/doc/html/boost_test/section_faq.html#boost_test.section_faq.is_boost_test_thread_safe

You should only use BOOST_REQUIRE_EQUAL on the main thread.

File: tests/two_queue_tests.cpp [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+   app->executor().post( priority::highest,app->executor().queue1(), [&]() { rslts[6]=seq_num; ++seq_num; } );
+   app->executor().post( priority::highest,app->executor().queue1(), [&]() { rslts[7]=seq_num; ++seq_num; } );
+   app->executor().post( priority::high,   app->executor().queue1(), [&]() { rslts[8]=seq_num; ++seq_num; } );
+   app->executor().post( priority::high,   app->executor().queue2(), [&]() { rslts[9]=seq_num; ++seq_num; } );
+
+   // stop application. Use lowest at the end to make sure this executes the last
+   app->executor().post( priority::lowest, app->executor().queue1(), [&]() {
+      // queue1 should have current function and queue2 should have all its functions
+      BOOST_REQUIRE_EQUAL( app->executor().queue1().size(), 1);
+      BOOST_REQUIRE_EQUAL( app->executor().queue2().size(), 4 );
```
Comment (heifner) [inline tests/two_queue_tests.cpp:88]: Same comment as above.

File: tests/two_queue_tests.cpp [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+   app->executor().post( priority::highest,app->executor().queue2(), [&]() { rslts[6]=seq_num; ++seq_num; } );
+   app->executor().post( priority::highest,app->executor().queue2(), [&]() { rslts[7]=seq_num; ++seq_num; } );
+   app->executor().post( priority::high,   app->executor().queue2(), [&]() { rslts[8]=seq_num; ++seq_num; } );
+   app->executor().post( priority::high,   app->executor().queue2(), [&]() { rslts[9]=seq_num; ++seq_num; } );
+
+   // Stop application. Use lowest at the end to make sure this executes the last
+   app->executor().post( priority::lowest, app->executor().queue1(), [&]() {
+      // queue1 should have current function and queue2 should have all its functions
+      BOOST_REQUIRE_EQUAL( app->executor().queue1().size(), 1);
+      BOOST_REQUIRE_EQUAL( app->executor().queue2().size(), 10 );
```
Comment (heifner) [inline tests/two_queue_tests.cpp:133]: Same comment as above.

File: tests/two_queue_tests.cpp [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+   app->executor().post( priority::lowest, app->executor().queue1(), [&]() { rslts[8]=seq_num; ++seq_num; } );
+   app->executor().post( priority::lowest, app->executor().queue1(), [&]() { rslts[9]=seq_num; ++seq_num; } );
+   app->executor().post( priority::low,    app->executor().queue2(), [&]() { rslts[10]=seq_num; ++seq_num; } );
+   app->executor().post( priority::medium, app->executor().queue2(), [&]() { rslts[11]=seq_num; ++seq_num; } );
+
+   // stop application. Use lowest at the end to make sure this executes the last
+   app->executor().post( priority::lowest, app->executor().queue1(), [&]() {
+      // queue1 should have current function and queue2's functions are all executed 
+      BOOST_REQUIRE_EQUAL( app->executor().queue1().size(), 1);
+      BOOST_REQUIRE_EQUAL( app->executor().queue2().size(), 0 );
```
Comment (heifner) [inline tests/two_queue_tests.cpp:175]: Same comment as above.

**Review by heifner — CHANGES_REQUESTED** [review]

File: tests/two_queue_tests.cpp (hunk 14 above)
Comment (linh2931) [reply tests/two_queue_tests.cpp:43]: Thanks. In this case when the two `BOOST_REQUIRE_EQUAL`s on the app thread are running,  the `BOOST_REQUIRE_EQUAL`s on the main thread have not started running yet. No conflicts would occur between the two threads.

File: tests/two_queue_tests.cpp (hunk 17 above)
Comment (greg7mdp) [reply tests/two_queue_tests.cpp:175]: These comments show why the two queue implementation should *not* be in appbase. It is specific to leap and unnecessarily add complexity to appbase.

**Review by heifner — APPROVED** [review]
